src/librustc_mir/interpret/eval_context.rs

```diff
@@ -48,8 +48,38 @@ pub struct InterpCx<'mir, 'tcx, M: Machine<'mir, 'tcx>> {
         FxHashMap<(Ty<'tcx>, Option<ty::PolyExistentialTraitRef<'tcx>>), Pointer<M::PointerTag>>,
 }
 
+struct SpanGuard(log::span::Span);
+
+impl SpanGuard {
+    fn new() -> Self {
+        Self(log::span::Span::none())
+    }
+
+    fn enter(&mut self, span: log::span::Span) {
+        if let Some(id) = span.id() {
+            log::dispatcher::get_default(|d| {
+                d.enter(&id);
+            });
+        }
+        // This executes the destructor on the previous instance of `SpanGuard`, ensuring that
+        // we never enter or exit more spans than vice versa. Unless you `mem::leak`, then we
+        // can't protect the tracing stack, but that'll just lead to weird logging, no actual
+        // problems.
+        *self = Self(span);
```

Comment:
Wow this is super confusing and IMO needs more comments.^^

So the thing we do here is that `SpanGuard` calls "enter" on something in `log::dispacher` inside `SpanGuard::enter`, and it calls `exit` on drop. But we call the exit *after* the enter, and somehow it still works? And what if we create a `SpanGaurd` and never `enter` it, then we `exit` something that was never entered?

It doesn't help that `Span` is already something very different in rustc so my brain just makes all the wrong connections all the time.^^

Reply:
Yea, the ordering wasn't ideal (I changed it), but also technically irrelevant as `enter` never gets called twice, but if it were, this would be sane.

I documented all this on the `new´ and `enter` methods